src/runtime/libuv.cpp

```diff
@@ -2,21 +2,267 @@
 Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 
-Author: Markus Himmel
-*/
+Author: Markus Himmel, Sofia Rodrigues
+ */
+#include <pthread.h>
+
 #include "runtime/libuv.h"
 
+#include "runtime/object.h"
+
 #ifndef LEAN_EMSCRIPTEN
 #include <uv.h>
 
+using namespace lean;
+
+uv_loop_t* global_uv_loop;
+uv_async_t* global_uv_signal_async;
+uv_mutex_t* global_uv_mutex;
+
+uv_cond_t* global_uv_cond;
+
+_Atomic(uint64_t) global_uv_n_waiters(0);
+
+static void lean_uv_wake() {
+    uv_async_send(global_uv_signal_async);
+}
+
+static void lean_uv_lock() {
+    if (uv_mutex_trylock(global_uv_mutex) != 0) {
+        global_uv_n_waiters++;
+        __sync_synchronize();
+        lean_uv_wake();
+
+        uv_mutex_lock(global_uv_mutex);
+
+        global_uv_n_waiters--;
+    }
+}
+
+static void lean_uv_unlock() {
+    if (uv_mutex_trylock(global_uv_mutex) != 0) {
+        uv_mutex_unlock(global_uv_mutex);
+    }
+
+    if (global_uv_n_waiters == 0) {
+        uv_cond_signal(global_uv_cond);
+    }
+}
+
+static void lean_uv_run() {
+    while (1) {
+        uv_mutex_lock(global_uv_mutex);
+
+        if (global_uv_n_waiters != 0) {
+            uv_cond_wait(global_uv_cond, global_uv_mutex);
+        }
+
+        if (global_uv_n_waiters == 0) {
+            uv_run(global_uv_loop, UV_RUN_ONCE);
+        }
+
+        uv_mutex_unlock(global_uv_mutex);
+    }
+}
+
+// Bindings
+
+static void lean_uv_timer_finalizer(void* ptr) {
+    lean_uv_timer_object* timer_obj = (lean_uv_timer_object*) ptr;
+
+    if (timer_obj->m_promise == NULL) {
+        /// The free can run after the lean_object is freed!
+        timer_obj->m_uv_timer.data = ptr;
+        uv_close((uv_handle_t*) & timer_obj->m_uv_timer, [](uv_handle_t* handle) {
```

Comment:
If I use `uv_mutex_lock` twice in the same thread the program will crash and we have two scenarios that the program can finalize the promise.

- The promise never called `next` and it will be finalized in the same thread, and this thread is not holding the mutex.
- The promise called `next` and will be finalized in the event loop thread, and this thread is holding the mutex.

Reply:
Changed to a recursive mutex.